src/glang/lib/std/rb_tree.c3

```diff
@@ -0,0 +1,273 @@
+@require_once "std/memory.c3"
+@require_once "std/wrappers.c3"
+
+function RED : () -> bool = {
+    return false;
+}
+
+function BLACK : () -> bool = {
+    return true;
+}
+
+function LEFT : () -> bool = {
+    return false;
+}
+
+function RIGHT : () -> bool = {
+    return true;
+}
+
+function left_if : (condition : bool) -> bool = {
+    if condition {
+        return LEFT();
+    }
+
+    return RIGHT();
+}
+
+typedef [K, V] RB_Node : {
+    key    : K,
+    value  : V,
```

Comment:
Interestingly llvm/ gcc's implementations don't have a key-value pair in each node, instead, they only store a single piece of data. So for example to implement a map you would specialize it as `RB_Node<MapData<K, V>>` (where `MapData = {key : K, value : V}`). This would also make ordered sets trivial `RB_Node<K>`.

For this to work, you'd need to adjust the interface slightly since we can't use the less-than comparison. Instead of `insert(tree, key, value)`, you could have `insert_in_position(tree, predecessor, data)` called by wrappers in `map` and `set` respectively. This has the interesting consequence that the red-black tree doesn't actually perform any comparisons of its nodes, it just preserves the ordering you give it... so maybe we could store something that can't be ordered with comparisons :scream:, not sure what tho :upside_down_face:.

Reply:
I think this is sensible, but we have an opportunity here to avoid a mistake (IMHO) in the C++ standard library and enable heterogeneous lookups by default. This lets us keep the `insert` function, but rely on an operator < template that we'll define, which will call the user-defined operator < for the key.

Not sure if there are any obvious downsides to this approach though